Implementation of the book addition form

kowalczyk256 opened the pull request:
Implementation of a simple form for adding books to the database.

kowalczyk256 committed:
Add HTMX template and basic adding form

kowalczyk256 committed:
Merge booksAPI into Book_Adding_Form

kowalczyk256 committed:
Improved add form

MrFranQx reviewed (commented):
There is a lot to unpack here.

First things first: `BookAdding` is NOT a good name. Remember, the name of the file is also the link. Something like plain `Add` would work better, in my opinion. And ideally, we should have `/Book/Add` and later `/Book/{id}/Edit`. Not sure if this is doable, though.

Now, onto the code:
- `<main>` is provided by layout, it should not be used twice.
- Populating `<select>`s dynamically is a good idea, but I would make them always visible and mark the empty ones as `disabled`.
- Validation and labelling need to be redone, see login and register forms as an example. And preferably check out #48 for that.
- Character count is also a cool idea, but IMO it needs to be somehow styled in-flow and not with hacky positioning. I have an idea in mind for that.
- And then, oh my, the confirmation dialog. It's a good idea, but it needs to be completely redone. Currently it doesn't even display properly in dark mode. See [Pico docs](https://picocss.com/docs/modal) for an idea.
- Formatting price doesn't need to be done with JavaScript, `<input type="number">` with a decimal `step` should suffice. And if not that, there's `pattern`.
- Not sure if all this validation of user login is necessary.
- On the backend we should construct a model like in login and register.
- Redirecting to the added book would be better than to the index page.

There's probably a lot more, but let's work gradually.
Of course, you don't have to agree on every one of these points. I'm open to discussion.

theluckyprogrammer commented on Booker/Pages/BookAdding.cshtml:
```diff
@@ -0,0 +1,227 @@
+                            hx-include="[name='Grade'], [name='Level']"
+                            data-next-select="grade">
+                        <option value="">Wybierz przedmiot</option>
+                        @foreach (var subject in Model.Subjects ?? new List<string>())
```
suggestion: ?? Enumerable.Empty<string>()

theluckyprogrammer commented on Booker/Pages/BookAdding.cshtml:
```diff
@@ -0,0 +1,227 @@
+                            class="hidden"
+                            data-next-select="level">
+                        <option value="">Wybierz klasę</option>
+                        @foreach (var grade in Model.Grades ?? new List<string>())
```
same here,
or you can create an extended method like:

```using System.Collections.Generic;
using System.Linq;

namespace YourProjectNamespace // Adjust the namespace accordingly
{
    public static class EnumerableExtensions
    {
        /// <summary>
        /// Returns the source enumerable, or an empty enumerable if the source is null.
        /// This prevents NullReferenceExceptions when iterating.
        /// </summary>
        /// <typeparam name="T">The type of elements in the enumerable.</typeparam>
        /// <param name="source">The source enumerable.</param>
        /// <returns>The source enumerable if not null; otherwise, an empty enumerable.</returns>
        public static IEnumerable<T> EmptyIfNull<T>(this IEnumerable<T>? source)
        {
            // If the source is null, return an empty sequence of the same type
            // Otherwise, return the original source
            return source ?? Enumerable.Empty<T>();
        }
    }
}
```
and then use it on the Model.Grades.EmptyIfNull() and any other fitting collection

theluckyprogrammer commented on Booker/Pages/BookAdding.cshtml.cs:
```diff
@@ -0,0 +1,173 @@
+
+        public async Task<IActionResult> OnGetAsync()
+        {
+            IsUserAuthenticated = User.Identity?.IsAuthenticated ?? false;
```
IsUserAuthenticated  <-- you really need that variable ? ;)

more over, you can create a extended method on User class and it will be doing the null check returning a value, and User object is everything you need in this class, it is static object after all.

theluckyprogrammer reviewed (dismissed):
In general, i see the pattern with polish label wandering from class to class, it need to be put to one place. This is proposition for another task

kowalczyk256 committed:
Merge branch 'booksAPI' into Book_Adding_Form

kowalczyk256 committed:
Another one version of adding books form - still need improvements

kowalczyk256 committed:
Merge branch 'booksAPI' into Book_Adding_Form

MrFranQx committed:
Yet another version, should be working now:

- utilize tag helpers and backend InputModel
- add automatic validation
- dynamic selects with HTMX partial
- basic caching
- correct connection string
- clean up and minimize JavaScript
- minor UI improvements

MrFranQx committed:
Fix styling

MrFranQx committed:
Remove redundant auth verification

MrFranQx commented:
I've done some work on that on my own. See `feature/add-form-improved`.

MrFranQx committed:
Fix redirect when unauthorized

MrFranQx committed:
Remove unneeded upload page

MrFranQx committed:
Merge main branch

theluckyprogrammer reviewed (approved):
In the future:
- Change IQueryable to IEnumerable to clarify intention (IQueryable is mostly connected with requesting data from the DB).
- Methods that don’t change object states may be placed in a separate file as extension methods.
- POST for adding a book: this may be a candidate for a stored procedure, or at least a candidate for a small refactor that will narrow down the connections to the database.
